File: apps/lib/services/localization_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter/material.dart';

/// Supported locales in Bay Navigator
/// Matches the web i18n configuration (10 languages)
enum AppLocale {
  en('en', 'English', 'English', '🇺🇸', false),
  es('es', 'Spanish', 'Español', '🇪🇸', false),
  zhHans('zh-Hans', 'Chinese (Simplified)', '简体中文', '🇨🇳', false),
  zhHant('zh-Hant', 'Chinese (Traditional)', '繁體中文', '🇹🇼', false),
  vi('vi', 'Vietnamese', 'Tiếng Việt', '🇻🇳', false),
  fil('fil', 'Filipino', 'Filipino', '🇵🇭', false),
  ko('ko', 'Korean', '한국어', '🇰🇷', false),
  ru('ru', 'Russian', 'Русский', '🇷🇺', false),
  fr('fr', 'French', 'Français', '🇫🇷', false),
  ar('ar', 'Arabic', 'العربية', '🇸🇦', true);

  final String code;
  final String name;
  final String nativeName;
  final String flag;
  final bool isRtl;

  const AppLocale(this.code, this.name, this.nativeName, this.flag, this.isRtl);

  /// Get Flutter Locale from AppLocale
  Locale get flutterLocale {
    if (code == 'zh-Hans') {
      return const Locale.fromSubtags(languageCode: 'zh', scriptCode: 'Hans');
    }
    if (code == 'zh-Hant') {
      return const Locale.fromSubtags(languageCode: 'zh', scriptCode: 'Hant');
    }
    return Locale(code);
  }

  /// Get AppLocale from language code
  static AppLocale fromCode(String code) {
    // Handle browser language code mappings
    final normalizedCode = code.toLowerCase();

    // Map common browser codes to our locale codes
    if (normalizedCode == 'zh-cn' || normalizedCode == 'zh-sg') {
      return AppLocale.zhHans;
    }
    if (normalizedCode == 'zh-tw' || normalizedCode == 'zh-hk' || normalizedCode == 'zh-mo') {
      return AppLocale.zhHant;
    }
    if (normalizedCode == 'tl') {
      return AppLocale.fil; // Tagalog -> Filipino
    }

    return AppLocale.values.firstWhere(
      (locale) => locale.code == code || locale.code.split('-')[0] == normalizedCode,
      orElse: () => AppLocale.en,
    );
  }
}

/// Localization service for Bay Navigator
/// Fetches translations from the web API and caches them locally
class LocalizationService {
  static const String _baseUrl = 'https://baynavigator.org/i18n/json';
  static const String _localeKey = 'baynavigator_locale';
  static const String _translationsCachePrefix = 'baynavigator_i18n_';
  static const Duration _cacheDuration = Duration(days: 7);

  final http.Client _client;
  SharedPreferences? _prefs;

  // In-memory cache of loaded translations
  final Map<String, Map<String, dynamic>> _translations = {};

  // Current locale
  AppLocale _currentLocale = AppLocale.en;

  LocalizationService({http.Client? client}) : _client = client ?? http.Client();

  AppLocale get currentLocale => _currentLocale;

  Future<SharedPreferences> get _preferences async {
    _prefs ??= await SharedPreferences.getInstance();
    return _prefs!;
  }

  /// Initialize the localization service
  /// Loads saved locale preference or detects from system
  Future<void> initialize() async {
    final prefs = await _preferences;

    // Check for saved locale preference
    final savedLocale = prefs.getString(_localeKey);
    if (savedLocale != null) {
      _currentLocale = AppLocale.fromCode(savedLocale);
    } else {
      // Use system locale if supported, otherwise default to English
      final systemLocale = WidgetsBinding.instance.platformDispatcher.locale;
      final systemLangCode = systemLocale.languageCode;

      // Check if system language is supported
      final matchingLocale = AppLocale.values.where((l) {
        if (l.code.contains('-')) {
          return l.code.split('-')[0] == systemLangCode;
        }
        return l.code == systemLangCode;
      }).firstOrNull;

      if (matchingLocale != null) {
        _currentLocale = matchingLocale;
      }
    }

    // Load translations for current locale
    await loadTranslations(_currentLocale);
  }

  /// Set the current locale and persist preference
  Future<void> setLocale(AppLocale locale) async {
    _currentLocale = locale;

    final prefs = await _preferences;
    await prefs.setString(_localeKey, locale.code);

    // Load translations if not already loaded
    await loadTranslations(locale);
  }

  /// Load translations for a locale
  /// First checks cache, then fetches from web if needed
  Future<void> loadTranslations(AppLocale locale) async {
    // Check if already loaded in memory
    if (_translations.containsKey(locale.code)) {
      return;
    }

    // Try to load from cache
    final cached = await _loadFromCache(locale.code);
    if (cached != null) {
      _translations[locale.code] = cached;
      return;
    }

    // Fetch from web
    await _fetchTranslations(locale.code);
  }

  /// Fetch translations from web API
  Future<void> _fetchTranslations(String localeCode) async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/$localeCode-ui.json'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        _translations[localeCode] = data;

        // Save to cache
        await _saveToCache(localeCode, data);
      }
    } catch (e) {
      // If fetch fails, try to use stale cache
      final staleCache = await _loadFromCache(localeCode, allowStale: true);
      if (staleCache != null) {
        _translations[localeCode] = staleCache;
      }
    }
  }

  /// Load translations from local cache
  Future<Map<String, dynamic>?> _loadFromCache(
    String localeCode, {
    bool allowStale = false,
  }) async {
    try {
      final prefs = await _preferences;
      final cached = prefs.getString('$_translationsCachePrefix$localeCode');
      if (cached == null) return null;

      final data = jsonDecode(cached) as Map<String, dynamic>;
      final timestamp = data['timestamp'] as int;
      final age = DateTime.now().millisecondsSinceEpoch - timestamp;

      if (!allowStale && age > _cacheDuration.inMilliseconds) {
        return null;
      }

      return data['translations'] as Map<String, dynamic>;
    } catch (e) {
      return null;
    }
  }

  /// Save translations to local cache
  Future<void> _saveToCache(String localeCode, Map<String, dynamic> data) async {
    try {
      final prefs = await _preferences;
      final cached = jsonEncode({
        'translations': data,
        'timestamp': DateTime.now().millisecondsSinceEpoch,
      });
      await prefs.setString('$_translationsCachePrefix$localeCode', cached);
    } catch (e) {
      // Ignore cache write errors
    }
  }

  /// Get a translation by key
  /// Supports dot notation (e.g., 'common.search')
  /// Falls back to English if translation not found
  String t(String key, [Map<String, dynamic>? params]) {
    // Try current locale first
    var value = _getNestedValue(_translations[_currentLocale.code], key);

    // Fall back to English if not found
    if (value == null && _currentLocale != AppLocale.en) {
      value = _getNestedValue(_translations[AppLocale.en.code], key);
    }

    // Return key if still not found
    if (value == null) {
      return key;
    }

    // Interpolate parameters
    if (params != null) {
      value = _interpolate(value, params);
    }

    return value;
  }

  /// Get nested value from map using dot notation
  String? _getNestedValue(Map<String, dynamic>? map, String key) {
    if (map == null) return null;

    final keys = key.split('.');
    dynamic current = map;

    for (final k in keys) {
      if (current is! Map<String, dynamic>) {
        return null;
      }
      current = current[k];
      if (current == null) {
        return null;
      }
    }

    return current is String ? current : null;
  }

  /// Interpolate parameters into a string
  /// Supports {param} syntax
  String _interpolate(String str, Map<String, dynamic> params) {
    var result = str;
    params.forEach((key, value) {
      result = result.replaceAll('{$key}', value.toString());
    });
    return result;
  }

  /// Clear all cached translations
  Future<void> clearCache() async {
    final prefs = await _preferences;
    for (final locale in AppLocale.values) {
      await prefs.remove('$_translationsCachePrefix${locale.code}');
    }
    _translations.clear();
  }

  /// Preload translations for all supported locales
  /// Useful for offline support
  Future<void> preloadAllTranslations() async {
    for (final locale in AppLocale.values) {
      await loadTranslations(locale);
    }
  }
}

File: apps/apple/BayNavigatorCore/Sources/BayNavigatorCore/Services/LocalizationService.swift
import Foundation

/// Supported locales in Bay Navigator
/// Matches the web and Flutter i18n configuration (10 languages)
public enum AppLocale: String, CaseIterable, Identifiable, Sendable {
    case en = "en"
    case es = "es"
    case zhHans = "zh-Hans"
    case zhHant = "zh-Hant"
    case vi = "vi"
    case fil = "fil"
    case ko = "ko"
    case ru = "ru"
    case fr = "fr"
    case ar = "ar"

    public var id: String { rawValue }

    public var name: String {
        switch self {
        case .en: return "English"
        case .es: return "Spanish"
        case .zhHans: return "Chinese (Simplified)"
        case .zhHant: return "Chinese (Traditional)"
        case .vi: return "Vietnamese"
        case .fil: return "Filipino"
        case .ko: return "Korean"
        case .ru: return "Russian"
        case .fr: return "French"
        case .ar: return "Arabic"
        }
    }

    public var nativeName: String {
        switch self {
        case .en: return "English"
        case .es: return "Espanol"
        case .zhHans: return "简体中文"
        case .zhHant: return "繁體中文"
        case .vi: return "Tieng Viet"
        case .fil: return "Filipino"
        case .ko: return "한국어"
        case .ru: return "Русский"
        case .fr: return "Francais"
        case .ar: return "العربية"
        }
    }

    public var flag: String {
        switch self {
        case .en: return "\u{1F1FA}\u{1F1F8}"
        case .es: return "\u{1F1EA}\u{1F1F8}"
        case .zhHans: return "\u{1F1E8}\u{1F1F3}"
        case .zhHant: return "\u{1F1F9}\u{1F1FC}"
        case .vi: return "\u{1F1FB}\u{1F1F3}"
        case .fil: return "\u{1F1F5}\u{1F1ED}"
        case .ko: return "\u{1F1F0}\u{1F1F7}"
        case .ru: return "\u{1F1F7}\u{1F1FA}"
        case .fr: return "\u{1F1EB}\u{1F1F7}"
        case .ar: return "\u{1F1F8}\u{1F1E6}"
        }
    }

    public var isRtl: Bool {
        self == .ar
    }

    /// Create from language code, handling common browser/system codes
    public static func from(code: String) -> AppLocale {
        let normalized = code.lowercased()

        // Map common system codes to our locale codes
        switch normalized {
        case "zh-cn", "zh-sg", "zh-hans":
            return .zhHans
        case "zh-tw", "zh-hk", "zh-mo", "zh-hant":
            return .zhHant
        case "tl": // Tagalog -> Filipino
            return .fil
        default:
            // Try direct match
            if let locale = AppLocale(rawValue: code) {
                return locale
            }
            // Try base language match
            let baseLang = normalized.split(separator: "-").first.map(String.init) ?? normalized
            return AppLocale.allCases.first { $0.rawValue.hasPrefix(baseLang) } ?? .en
        }
    }
}

/// Localization service for Bay Navigator apps
/// Fetches translations from the web API and caches them locally
public actor LocalizationService {
    public static let shared = LocalizationService()

    private let baseUrl = "https://baynavigator.org/i18n/json"
    private let localeKey = "baynavigator:locale"
    private let translationsCachePrefix = "baynavigator:i18n_"
    private let cacheDuration: TimeInterval = 7 * 24 * 60 * 60 // 7 days

    private let defaults = UserDefaults.standard

    // In-memory cache of loaded translations
    private var translations: [String: [String: Any]] = [:]

    // Current locale
    public private(set) var currentLocale: AppLocale = .en

    private init() {}

    // MARK: - Initialization

    /// Initialize the localization service
    /// Loads saved locale preference or detects from system
    public func initialize() async {
        // Check for saved locale preference
        if let savedCode = defaults.string(forKey: localeKey) {
            currentLocale = AppLocale.from(code: savedCode)
        } else {
            // Use system locale if supported
            let preferredLanguages = Locale.preferredLanguages
            for lang in preferredLanguages {
                let locale = AppLocale.from(code: lang)
                if locale != .en || lang.hasPrefix("en") {
                    currentLocale = locale
                    break
                }
            }
        }

        // Load translations for current locale
        await loadTranslations(for: currentLocale)
    }

    // MARK: - Locale Management

    /// Set the current locale and persist preference
    public func setLocale(_ locale: AppLocale) async {
        currentLocale = locale
        defaults.set(locale.rawValue, forKey: localeKey)

        // Load translations if not already loaded
        await loadTranslations(for: locale)
    }

    /// Get the current locale (for non-async contexts)
    nonisolated public func getCurrentLocale() -> AppLocale {
        if let savedCode = UserDefaults.standard.string(forKey: "baynavigator:locale") {
            return AppLocale.from(code: savedCode)
        }
        return .en
    }

    // MARK: - Translation Loading

    /// Load translations for a locale
    /// First checks cache, then fetches from web if needed
    public func loadTranslations(for locale: AppLocale) async {
        // Check if already loaded in memory
        if translations[locale.rawValue] != nil {
            return
        }

        // Try to load from cache
        if let cached = loadFromCache(locale.rawValue) {
            translations[locale.rawValue] = cached
            return
        }

        // Fetch from web
        await fetchTranslations(locale.rawValue)
    }

    /// Fetch translations from web API
    private func fetchTranslations(_ localeCode: String) async {
        guard let url = URL(string: "\(baseUrl)/\(localeCode)-ui.json") else {
            return
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                // Try stale cache on error
                if let stale = loadFromCache(localeCode, allowStale: true) {
                    translations[localeCode] = stale
                }
                return
            }

            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                translations[localeCode] = json
                saveToCache(localeCode, data: json)
            }
        } catch {
            // Try stale cache on error
            if let stale = loadFromCache(localeCode, allowStale: true) {
                translations[localeCode] = stale
            }
        }
    }

    // MARK: - Cache Management

    private struct CachedTranslations: Codable {
        let timestamp: TimeInterval
        let data: Data
    }

    private func loadFromCache(_ localeCode: String, allowStale: Bool = false) -> [String: Any]? {
        guard let data = defaults.data(forKey: "\(translationsCachePrefix)\(localeCode)") else {
            return nil
        }

        do {
            let cached = try JSONDecoder().decode(CachedTranslations.self, from: data)
            let age = Date().timeIntervalSince1970 - cached.timestamp

            if !allowStale && age > cacheDuration {
                return nil
            }

            return try JSONSerialization.jsonObject(with: cached.data) as? [String: Any]
        } catch {
            return nil
        }
    }

    private func saveToCache(_ localeCode: String, data: [String: Any]) {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: data)
            let cached = CachedTranslations(
                timestamp: Date().timeIntervalSince1970,
                data: jsonData
            )
            let encodedData = try JSONEncoder().encode(cached)
            defaults.set(encodedData, forKey: "\(translationsCachePrefix)\(localeCode)")
        } catch {
            // Ignore cache write errors
        }
    }

    /// Clear all cached translations
    public func clearCache() {
        for locale in AppLocale.allCases {
            defaults.removeObject(forKey: "\(translationsCachePrefix)\(locale.rawValue)")
        }
        translations.removeAll()
    }

    // MARK: - Translation Lookup

    /// Get a translation by key
    /// Supports dot notation (e.g., "common.search")
    /// Falls back to English if translation not found
    public func t(_ key: String, params: [String: Any]? = nil) -> String {
        // Try current locale first
        var value = getNestedValue(translations[currentLocale.rawValue], key: key)

        // Fall back to English if not found
        if value == nil && currentLocale != .en {
            value = getNestedValue(translations[AppLocale.en.rawValue], key: key)
        }

        // Return key if still not found
        guard var result = value else {
            return key
        }

        // Interpolate parameters
        if let params = params {
            for (paramKey, paramValue) in params {
                result = result.replacingOccurrences(of: "{\(paramKey)}", with: "\(paramValue)")
            }
        }

        return result
    }

    /// Get nested value from dictionary using dot notation
    private func getNestedValue(_ dict: [String: Any]?, key: String) -> String? {
        guard let dict = dict else { return nil }

        let keys = key.split(separator: ".").map(String.init)
        var current: Any = dict

        for k in keys {
            guard let currentDict = current as? [String: Any],
                  let next = currentDict[k] else {
                return nil
            }
            current = next
        }

        return current as? String
    }

    /// Preload translations for all supported locales
    /// Useful for offline support
    public func preloadAllTranslations() async {
        for locale in AppLocale.allCases {
            await loadTranslations(for: locale)
        }
    }
}
